lib/src/source/hosted.dart

```diff
@@ -1100,3 +1138,83 @@ class _RefAndCache {
   @override
   bool operator ==(Object other) => other is _RefAndCache && other.ref == ref;
 }
+
+const checksumHeaderName = 'x-goog-hash';
+
+/// Adds a checksum validation "tap" to the response stream and returns a
+/// wrapped `Stream` object, which should be used to consume the incoming data.
+///
+/// As chunks are received, a CRC32C checksum is updated.
+/// Once the download is completed, the final checksum is compared with
+/// the one present in the checksum response header.
+///
+/// Throws [PackageIntegrityException] if there is a checksum mismatch.
+Stream<List<int>> _validateStream(
+    Stream<List<int>> stream, int expectedChecksum, String fileName) {
+  return Crc32c.computeByTappingStream(stream, handleDone: (actualChecksum) {
+    log.fine(
+        'Computed checksum $actualChecksum for "$fileName" with expected CRC32C of $expectedChecksum');
+
+    if (actualChecksum != expectedChecksum) {
+      throw PackageIntegrityException(
+        'Package archive "$fileName" has a CRC32C checksum mismatch',
+      );
+    }
+  });
+}
+
+/// Parses response [headers] and returns the archive's CRC32C checksum.
+///
+/// In most cases, GCS provides both MD5 and CRC32C checksums in its response
+/// headers. It uses the header name "x-goog-hash" for these values. It has
+/// been documented and observed that GCS will send multiple response headers
+/// with the same "x-goog-hash" token as the key.
+/// https://cloud.google.com/storage/docs/xml-api/reference-headers#xgooghash
+///
+/// Additionally, when the Dart http client encounters multiple response
+/// headers with the same key, it concatenates their values with a comma
+/// before inserting a single item with that key and concatenated value into
+/// its response "headers" Map.
+/// See https://github.com/dart-lang/http/issues/24
+/// https://github.com/dart-lang/http/blob/06649afbb5847dbb0293816ba8348766b116e419/pkgs/http/lib/src/base_response.dart#L29
+///
+/// Throws [PackageIntegrityException] if the CRC32C checksum cannot be parsed.
+int? _parseCrc32c(Map<String, String> headers, String fileName) {
+  final checksumHeader = headers[checksumHeaderName];
+  if (checksumHeader == null) return null;
+
+  final parts = checksumHeader.split(',');
+  for (final part in parts) {
+    if (part.startsWith('crc32c=')) {
+      final undecoded = part.substring('crc32c='.length);
+
+      try {
+        final bytes = base64.decode(undecoded);
+        return bytesToUint32(bytes);
+
+        // Suppress lint because 1) we are dealing with arbitrary runtime values
+        // 2) we do not want to duplicate the validation logic that is already
+        // throwing the `RangeError`.
+        // ignore: avoid_catching_errors
+      } on RangeError {
+        continue;
+      } on FormatException {
+        continue;
+      }
+    }
+  }
+
+  throw PackageIntegrityException(
+    'Package archive "$fileName" has a malformed CRC32C checksum in its response headers',
```

Comment:
We could also ignore the case where the header is there, but there is no crc32c.
It feels safer to ignore, but I don't know of anyone that would make a server which returns an x-goog-hash header without crc32c in the header. But it seems like someone else emulating GCS might do that?

Reply:
Sure, I can make that change. I think that seems like straightforward reasoning.